Core/src/Geometry/GeometryIdGenerators.cpp

```diff
@@ -0,0 +1,178 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2016-2018 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Geometry/detail/GeometryIdGenerators.hpp"
+
+#include "Acts/Geometry/Portal.hpp"
+#include "Acts/Surfaces/Surface.hpp"
+#include "Acts/Utilities/Enumerate.hpp"
+
+#include <sstream>
+
+void Acts::Experimental::detail::VolumeCounter::generateIds(
+    DetectorVolume& volume) {
+  // The geometry Id has already been set, hence ignore
+  if (volume.geometryId().volume() != 0u) {
+    return;
+  }
+
+  GeometryIdentifier volumeID;
+  volumeID.setVolume(++volumeCounter);
+  volume.assignGeometryId(volumeID);
+
+  if (processSurfacesPortals) {
+    // Set to all contained surfaces
+    size_t sensitiveCounter = 0;
+    size_t passiveCounter = 0;
+    for (auto& sPtr : volume.surfacePtrs()) {
+      auto surfaceID = sPtr->geometryId();
+      if (sPtr->associatedDetectorElement() != nullptr and
+          surfaceID.sensitive() == 0u) {
+        sPtr->assignGeometryId(
+            (GeometryIdentifier(volumeID).setSensitive(++sensitiveCounter)));
+      } else if (surfaceID.passive() == 0u and
+                 sPtr->associatedDetectorElement() == nullptr) {
+        sPtr->assignGeometryId(
+            (GeometryIdentifier(volumeID).setPassive(++passiveCounter)));
+      }
+    }
```

Comment:
could we reuse the `SensitiveCounter` and `PassiveCounter` here?

Reply:
Right you are.